File: estudo-aeds-2/exercicios/arvoreVerde.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Node{
    int elemento;
    struct Node *esq, *dir;
}Node;

typedef struct Arvore{
    int elemento;
    Node *raiz;
}Arvore;

Arvore *criar_arvore(){
    Arvore *arvore = (Arvore *)malloc(sizeof(Arvore));
    arvore->raiz = NULL;

    return arvore;
}
void inserir(Arvore *a, int x)
{ 
    Node *novo = (Node *)malloc(sizeof(Node));
    novo->elemento = x;
    novo->esq = NULL;
    novo->dir = NULL;

    if(a->raiz == NULL){
        a->raiz = novo;
        return;
    }

    Node *pai = NULL;
    Node *atual = a->raiz;

    while(atual != NULL){
        if(atual->elemento > x){
            pai = atual;
            atual = atual->esq;
        }else{
            pai = atual;
            atual = atual->dir;
        }
    }
    if(pai->elemento > x)
        pai->esq = novo;
    else
        pai->dir = novo;
}
void pesquisar(Arvore *a, int y)
{
    if(a->raiz == NULL){
        printf("N\n");
        return;
    }

    Node *atual = a->raiz;
    atual = a->raiz;

    while(atual != NULL && atual->elemento != y){
        if(atual->elemento > y){
            printf("%d ", atual->elemento);
            atual = atual->esq;
        }else{
            printf("%d ", atual->elemento);
            atual = atual->dir;
        }
    }
    if(atual == NULL)
        printf("N\n");
    else
        printf("%d S\n", atual->elemento);
}
void caminharPre(Node *i)
{
    if(i != NULL){
        printf("%d " , i->elemento);
        caminharPre(i->esq);
        caminharPre(i->dir);
    }
}
void caminharPos(Node *i)
{
    if(i != NULL){
        caminharPos(i->esq);
        caminharPos(i->dir);
        printf("%d ", i->elemento);
    }
}
void caminharOrdem(Node *i)
{   
    if(i != NULL){
        caminharOrdem(i->esq);
        printf("%d " , i->elemento);
        caminharOrdem(i->dir);
    }
}
int main()
{
    Arvore *arvore = criar_arvore();
    int valor;
    char *comando = (char *)malloc(10 * sizeof(char));


    while(scanf("%s", comando) != EOF){
        if(strcmp(comando, "I") == 0){
            scanf("%d" , &valor);
            inserir(arvore, valor);
        }else if(strcmp(comando, "P") == 0){
            scanf("%d" , &valor);
            pesquisar(arvore, valor);
        }else if(strcmp(comando, "PRE") == 0){
            if(arvore->raiz == NULL) 
                printf("V\n");
            else{
                caminharPre(arvore->raiz);
                printf("\n");
            }
        }else if(strcmp(comando, "POS") == 0){
            if(arvore->raiz == NULL)
                printf("V\n");
            else{
                caminharPos(arvore->raiz);
                printf("\n");
            }
        }else if(strcmp(comando, "EM") == 0){
            if(arvore->raiz == NULL)
                printf("V\n");
            else{
                caminharOrdem(arvore->raiz);
                printf("\n");
            }
        }
    }
}

File: estudo-aeds-2/exercicios/ArvoreVerde.java
import java.util.*;
class Node
{
    int elemento;
    Node esq, dir;

    public Node(int x){
        this.elemento = x;
        this.esq = null;
        this.dir = null;
    }
}
class Arvore
{
    Node raiz;

    public Arvore(){
        raiz = null;
    }

    public void inserir(int x){
        Node tmp = new Node(x);

        if(raiz == null){
            raiz = tmp;
            return;
        }

        Node pai = null;
        Node atual = raiz;

        while(atual != null){
            if(atual.elemento > x){
                pai = atual;
                atual = atual.esq;
            }else{
                pai = atual;
                atual = atual.dir;
            }
        }

        if(pai.elemento > x)
            pai.esq = tmp;
        else
            pai.dir = tmp;
    }
    public void pesquisar(int y){
        if(raiz == null){
           System.out.println("N");
            return;
        }

        Node atual = raiz;

        while(atual != null &&  atual.elemento != y){
            if(atual.elemento > y){
                System.out.print(atual.elemento + " ");
                atual = atual.esq;
            }
            else{
                System.out.print(atual.elemento + " ");
                atual = atual.dir;
            }
        }
        if(atual == null)
            System.out.println("N");
        else
            System.out.println(atual.elemento + " S");
    }

    public void caminharPre(Node i){
        if(i != null){
            System.out.print(i.elemento + " ");
            caminharPre(i.esq);
            caminharPre(i.dir);
        }
    }
    public void caminharPos(Node i){
        if(i != null){
            caminharPos(i.esq);
            caminharPos(i.dir);
            System.out.print(i.elemento + " ");
        }
    }
    public void caminharOrdem(Node i){
        if(i != null){
            caminharOrdem(i.esq);
            System.out.print(i.elemento + " ");
            caminharOrdem(i.dir);
        }
     }
}

public class ArvoreVerde
{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        Arvore arvore = new Arvore();
        String comando;
        int valor;

        while(sc.hasNext()){
            comando = sc.next();
            if(comando.compareTo("I") == 0){
                valor = sc.nextInt();
                arvore.inserir(valor);
            }else if(comando.compareTo("P") == 0){
                valor = sc.nextInt();
                arvore.pesquisar(valor);
            }else if(comando.compareTo("PRE") == 0){
                if(arvore.raiz == null)
                    System.out.println("V");
                else{
                    arvore.caminharPre(arvore.raiz);
                    System.out.println();
                }
            }else if(comando.compareTo("POS") == 0){
                if(arvore.raiz == null)
                    System.out.println("V");
                else{
                    arvore.caminharPos(arvore.raiz);
                    System.out.println();
                }
            }else if(comando.compareTo("EM") == 0){
                if(arvore.raiz == null)
                    System.out.println("V");
                else{
                    arvore.caminharOrdem(arvore.raiz);
                    System.out.println();
                }
            }
        }
        sc.close();
    }
}
